mitodl/ol-data-platform
# 737 Log Normalization Dagster Pipeline

# What are the relevant tickets?
Closes [#737](https://github.com/mitodl/ol-data-platform/issues/737)

# Description (What does it do?)
Added a new Dagster pipeline to normalize historical tracking log data.
This pipeline takes in a bucket and start date parameters and then processes all the log files for that bucket starting at the start date bucket directory working down to the latest logs. The pipeline processes one date at a time and uses a glob pattern to load all the files for that date directly to a table. DuckDB loads the gzipped newline delimited json log files directly to an in-memory SQL engine, auto detecting the tracking log file data types while only parsing root level JSON. Updates are run to cast the context, event, and time fields from JSON to strings and integer timestamps are normalized to ISO8601. Once Complete, the pipeline writes the normalized data back to the S3 bucket, under the 'valid/' subdirectory, writing one normalized file per source file to match the original file names and structure.


# How can this be tested?
Ideally, these changes should be run locally with dagit and then deployed to our QA Dagster instance. Once the config_schema is finalized, we need to write a template to generate that in the AMI build to handle the IAM credentials to allow the pipeline to read from an write to S3.

Core logic may be tested using the manually run utility script or running the DuckDB SQL via DuckDB CLI.


## Checklist:
- [x] Address commit hook failures for debug statements, black, ruff, and mypy
- [x] Properly import/configure a duckDB resource
- [x] Finalize the config_schema and IAM credentials needed
- [ ] Create a job definition for each of the deployments to avoid having to run multiple containers
